GB3 homework 6

emikhailow opened the pull request:
GB3 homework 6

emikhailow committed:
GB3 homework 6

emgusev commented on src/main/java/homework3/ClientHandler.java:
```diff
@@ -63,7 +63,7 @@ public ClientHandler(MyServer server, Socket socket) {
         } catch (IOException e) {
 
-            System.out.println("Problem while creating client");
+            server.getLOGGER().error("[Server]: problem while creating client");
```
Логгер лучше делать свой в каждом классе. Тогда будет понятно какой класс и откуда пишет сообщение.  Сервер пишет свое. Клиент хэндлер свое

emgusev commented on src/main/java/homework6/Homework6App.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Homework6App {
+
+    public static ArrayList<Integer> ExtractSubArrayAfterLastOccurrence(List<Integer> arrayList, int occurence){
```
`public static List<Integer> extractSubArrayAfterLastOccurrence`

1) Интерфейс
2) название метода с маленькой буквы

emikhailow replied:
Справедливо, каюсь

emgusev commented on src/main/java/homework6/Homework6App.java:
```diff
@@ -0,0 +1,29 @@
+        }
+    }
+
+    public static boolean CheckArrayForContainingOnlyFollowingNumbers(List<Integer> list, Integer... ints){
```
с маленькой буквы) 

А вообще логика не по заданию.  Должен содержать ТОЛЬКО цифры 1 и 4 и обязательно обе
[1,4,3] - false 
[1] - false
[1,4] true

emikhailow replied:
Вроде как метод работает корректно у меня
При каких значениях он неправильно отработает?

emgusev replied:
Ниже комментарий написал :) Увидел что работает. Но оч неоптимально

emgusev commented on src/test/java/homework6/Homework6TestApp.java:
```diff
@@ -0,0 +1,61 @@
+    @Test
+    @DisplayName("ExtractSubArrayAfterLastOccurrence: as expected")
+    void testAsExpectedExtractSubArrayAfterLastOccurrence(){
+        HashMap<List<Integer>, List<Integer>> hashMap = new HashMap<>();
```
А параметризованный тест делает тоже самое :)

emikhailow replied:
Справедливо, поправлю

emgusev commented on src/test/java/homework6/Homework6TestApp.java:
```diff
@@ -0,0 +1,61 @@
+    @Test
+    @DisplayName("CheckArrayForContainingOnlyFollowingNumbers: as expected")
+    void testAsExpectedCheckArrayForContainingOnlyFollowingNumbers(){
+        HashMap<List<Integer>, Boolean> hashMap = new HashMap<>();
```
`Map<List<Integer>, Boolean> hashMap = new HashMap<>();` :)

emikhailow replied:
Справедливо, каюсь, исправил

emgusev commented on src/test/java/homework6/Homework6TestApp.java:
```diff
@@ -0,0 +1,61 @@
+        hashMap.put(List.of(1, 1, 1, 4, 4, 1, 4, 4), true);
+        hashMap.put(List.of(1, 1, 1, 1, 1, 1), true);
+        hashMap.put(List.of(4, 4, 4, 4), false);
+        hashMap.put(List.of(1, 4, 4, 1, 1, 4, 3), false);
```
Увидел что у тебя этот тест проходит. Даже понял почему. Но логика очень сложная. Ты 2 раза сравниваешь каждого с каждым. На большом размере массива будет тормозить :(

emikhailow replied:
А можем обсудить, как надо правильно?

emgusev replied:
2 флага типа boolean. есть ли 1 и есть ли 4

Итерируемся по массиву. если встретили НЕ 1 или 4, то return false;

иначе проверяем 2 флажка. 

Не теряем false для пустого массива


Получается всего один проход по массиву. сложность n, а не n*n

emikhailow commented on src/main/java/homework6/Homework6App.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public static boolean CheckArrayForContainingOnlyFollowingNumbers(List<Integer> list, Integer... ints){
+       return Arrays.asList(ints).containsAll(list) && list.containsAll(Arrays.asList(ints));
```
@emgusev, а можно чуть-чуть поспорить с тобой? 
я понимаю, что можно обойтись двумя флага, равно1 и равно4, но я попытался сделать общий случай, то есть, чтобы цифры, наличие которых нужно проверять в массиве, задавались через varargs. 
Сложность каждого containsAll это О(n * m), где n - длина массива, a m - количество вот этих цифр из varargs. Причем m скорее всего маленький, можно считать константа. Получается, что сложность О(2 * n * m) или O (n * m) или O(n). Можешь сказать, в чем я не прав? Я понимаю, что можно при помощи двух флагов задачу решить, но это как-то неуниверсально. Или ок?

emgusev replied:
Универсально то да. Но ContainsAll в любом случае ведет к просмотру всего массива. Дважды. при маленьком m ты действительно свалишься к 2 * m * n, но зачем, когда можно просто n? :) 

Хочется универсальности - ок :) Давай сделаем Map <Integer, Boolean> которая будет хранить признак присутствия чиселки. Мапа будет размера m. И сложность свалится к m+n

emikhailow committed:
GB3 homework 6
учел замечания по параметризированным тестам, по сложности прохода массива и по косметическим ошибкам
